# Data Analysis Project: Ads and Apps

## About: 
This is a Data Analytics project that is meant to answer questions about what apps are the ones that attract more users. The information comes from datasets tha can be found here:

[Google Dataset](https://www.kaggle.com/lava18/google-play-store-apps)

[Apple Dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

It is recommended to visit the links above to get a better idea of what each column of the datasets means.



In [1]:
from csv import reader

# AppleStore Dataset
file = open('C:/Users/david/Desktop/Autodidacta/Datasets/AppleStore.csv', encoding="utf8")
read_file = reader(file)
applestore_apps = list(read_file)
applestore_header = applestore_apps[0]
applestore = applestore_apps[1:]

#Google Play Dataset
file = open('C:/Users/david/Desktop/Autodidacta/Datasets/googleplaystore.csv', encoding="utf8")
read_file = reader(file)
google_apps = list(read_file)
google_header = google_apps[0]
google = google_apps[1:]




In [2]:
# Exploration of the lists with the data in a more readable way
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(row))

# Print the first rows, the number of rows and number of columns
explore_data(applestore, 0, 4, True)
explore_data(google, 0, 4, True)

#Columns Names
print("Apple Columns: \n")
print(applestore_header)
print("\n")
print("Google Columns: \n")
print(google_header)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live 

For out purpose the columns that can be of use are:

For Apple:
* track_name
* currency
* price
* rating_count_tot
* rating_count_tot_ver
* prime_genre


For Google:
* App
* Category
* Reviews
* Genres
* Installs

In the [discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion) of the google data set a row with an empty column is reported in position 10472 (With no header). There are come duplicates aswell.

The apple dataset apparently has no reported errors. This can be checked in the [discussion](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion) for this data set.

In [3]:
# Comparation of a reported row with error
print(google[10472])
print(google[10473])

# Delete the row with error
del google[10472]


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


The Google Play Apps dataset has many duplicate entries, so it is ideal to get rid of those. The main difference between reapeted entries is the number of reviews, so if we want the most recent entry we are going to want the one with most reviews.

In [4]:
#Count the number of duplicates
unique_entries = []
repeated_entries = []
for entry in google:
    name = entry[0]
    if name in unique_entries:
        repeated_entries.append(name)
    else:
        unique_entries.append(name)
print(len(repeated_entries))

1181


So, now we know there are 1181 duplicates, we want to remove them. This won't be a random operation, because, as we know from before we want the newest entry of the duplicate, so this will be our criterion of deletion.

In [5]:
#Find the highest reviews for each app
apps_reviews = {}

for entry in google:
    name = entry[0]
    review = float(entry[3])
    if name in apps_reviews and review > apps_reviews[name]:
        apps_reviews[name] = review
    elif name not in apps_reviews:
        apps_reviews[name] = review
#To see if the number of elements saved is the same than the length of the data set - 1181
print(len(apps_reviews))
print(len(google) - 1181)

9659
9659


The method used for the cleaning is having 2 lists, one that will store the rows with the highest reviews and the other wiht the names of the apps already added, this for the cases in which the app is repeated with the highest reviews number. Then we loop inside the google dataset and add the row to clean_google and name to added only if the value of reviews is the max one and the name of the app is not in added. 

In [6]:
clean_google = []
added = []
for app in google:
    name = app[0]
    reviews = float(app[3])
    if reviews == apps_reviews[name] and name not in added:
        clean_google.append(app)
        added.append(name)




Let's check the resultant list of lists

In [7]:
explore_data(clean_google, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows: 9659
Number of columns: 13


Next, we have to consider that this analysis is for apps that are for English speakers and there are apps that are for other languages that are not in our apps scope, so we shall remove them. Let's write a function for character detection.

What this function will do is:

Detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

In [8]:
def detect_characters(string):
    for character in string:
        if ord(character) > 127:
            return False
    return True

This works because it tests if a string has characters greater than 127 which is out of the range of numbers assigned to characters that are between the range for English speakers common characters.Let's test the function with the strings: 

* 'Instagram'
* '爱奇艺PPS -《欢乐颂2》电视剧热播'
* 'Docs To Go™ Free Office Suite'
* 'Instachat 😜'

In [9]:
print(detect_characters('Instagram'))
print(detect_characters('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(detect_characters('Docs To Go™ Free Office Suite'))
print(detect_characters('Instachat 😜'))


True
False
False
False


Som English apps use characters that are outside the range, like emojis, so we have to minimize this loss because these are apps that have value for us. Let's say that we can accept apps with 3 or less characters outside the range and change our function.

In [10]:
def detect_characters(string):
    ascii_counter = 0
    for character in string:
        if ord(character) > 127:
            ascii_counter += 1
    if ascii_counter > 3:
        return False
    else:
        return True

Let's test with the same strings

In [11]:
print(detect_characters('Instagram'))
print(detect_characters('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(detect_characters('Docs To Go™ Free Office Suite'))
print(detect_characters('Instachat 😜'))

True
False
True
True


With this we mitigated the loss of useful apps. Now let's use our new function to filter the datasets



In [12]:
#Filter the clean_google dataset
english_google = []
for app in clean_google:
    name = app[0]
    if detect_characters(name):
        english_google.append(app)

#Filter the applestore data set
english_apple = []
for app in applestore:
    name = app[0]
    if detect_characters(name):
        english_apple.append(app)


Let's explore the resulting data sets and see the remaining apps.


In [13]:
explore_data(english_google, 0, 3, True)
explore_data(english_apple, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'G

Now we are going to isolate only the free apps as our last data cleaning step.


In [14]:
free_google = []
for app in english_google:
    price = app[7]
    if price == '0':
        free_google.append(app)

free_apple = []
for app in english_apple:
    price = app[4]
    if price == '0.0':
        free_apple.append(app)
print(len(free_google))
print(len(free_apple))


8864
4056


Now that we:
* Removed inaccurate data
* Removed duplicate app entries
* Removed non-English apps
* Isolated the free apps

Our aim is to determine the kinds of apps that are likely to attract more users. First we want to find app profiles that are succesful both in the AppleStore and Google Play so let's begin the analysis by getting a sense of what are the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our data sets.

Let's create some functions that will minimize the work of creating the frequency tables and analyzing them.

In [15]:
#Function for getting a frequency table with percentages
def create_freqtable(dataset, index):
    freq_table = {}
    for app in dataset:
        column = app[index]
        if column in freq_table:
            freq_table[column] += 1
        else:
            freq_table[column] = 1
    percentages = {}        
    for key in freq_table:
        percentages[key] = (freq_table[key]/len(dataset))*100
    return percentages

#Function to sort frequency tables
def display_table(dataset, index):
    table = create_freqtable(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        

Now let's try our function for the columns Genre, Category and prime_genre and make analysis. Is important to remember our scope, which is free english-speaking apps, so the information analyzed here can not be generalized for all the App Store or Google play.

In [16]:
print('prime_genre Column')
display_table(free_apple, 11)


prime_genre Column
Games : 55.64595660749507
Entertainment : 8.234714003944774
Photo & Video : 4.117357001972387
Social Networking : 3.5256410256410255
Education : 3.2544378698224854
Shopping : 2.983234714003945
Utilities : 2.687376725838264
Lifestyle : 2.3175542406311638
Finance : 2.0710059171597637
Sports : 1.947731755424063
Health & Fitness : 1.8737672583826428
Music : 1.6518737672583828
Book : 1.6272189349112427
Productivity : 1.5285996055226825
News : 1.4299802761341223
Travel : 1.3806706114398422
Food & Drink : 1.0601577909270217
Weather : 0.7642998027613412
Reference : 0.4930966469428008
Navigation : 0.4930966469428008
Business : 0.4930966469428008
Catalogs : 0.22189349112426035
Medical : 0.19723865877712032


We are going to answer the next questions for the AppStore:
* What is the most common genre? What is the runner-up?
As we can see the most common genre in the appstore is Games with a grand 55.64% of the apps. The runner up would be Entertainment apps with 8.23%. 
* What is the general impression — are most of the apps designed for practical purposes (education, shopping, utilities, productivity, lifestyle) or more for entertainment (games, photo and video, social networking, sports, music)?
If we add up the percentages from each category we can see that the great majority of apps are designed for entertainment purposes while the minority is for practical purposes. 
* Can you recommend an app profile for the App Store market based on this frequency table alone? If there's a large number of apps for a particular genre, does that also imply that apps of that genre generally have a large number of users?
So summarizing, the majority of app offer in the AppStore is for fun, but this says only that there is great offer of this kind of app, but we don't know yet what is the demand of the users, therefore we can't know with this information which type of app has the most users.

Now let's analyze the information about google Play.


In [17]:
print('Category Column')
display_table(free_google, 1)


Category Column
FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
A

In [18]:
print('Genre Column')
display_table(free_google, 9)

Genre Column
Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.

For Google Play let's answer this questions:

* What are the most common genres?
Apparently the category of apps that has the higher percentage is Family with 18.90 % of the english-speaking free apps, followed up by game, tools, bussines and life style.
* Compare the patterns you see for the Google Play market with those you saw for the App Store market.
The situation between the AppStore and Play Store is very different. While in the AppStore apps designed for fun have the biggest numbers, in the Play Store practical purpose apps are a majority as a whole, something that can be confirmed with the leading genres which are tools, education, business, education, productivity, from which we can conclude that in the playstore english-speaking free apps the majority are of practical purpose, but the difference is not that great between the two groups because the family category is conformed by many games for kids.
* Can you recommend an app profile based on what you found so far? Do the frequency tables you generated reveal the most frequent app genres or what genres have the most users?
So a general landscape is that in the AppleStore the apps are mainly for fun and in the play store the practical purpose apps have the lead but it is more balanced in terms of the difference between this two classes of apps. 

Now to know what are the apps with most users we need to make more analysis.

For the AppStore we are going to use the total ratings of an app to define the total retings for each genre to get an idea about what are the most popular genres inside the AppStore

In [21]:
prime_genere_ft = create_freqtable(free_apple, 11)
for genre in prime_genere_ft:
    total = 0
    len_genre = 0
    for app in free_apple:
        genre_app = app[11]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_number_ratings = total/len_genre
    print('Genre: ' + genre)
    print('Average Number of User Ratings: ' + str(avg_number_ratings))

Genre: Social Networking
Average Number of User Ratings: 53078.195804195806
Genre: Photo & Video
Average Number of User Ratings: 27249.892215568863
Genre: Games
Average Number of User Ratings: 18924.68896765618
Genre: Music
Average Number of User Ratings: 56482.02985074627
Genre: Reference
Average Number of User Ratings: 67447.9
Genre: Health & Fitness
Average Number of User Ratings: 19952.315789473683
Genre: Weather
Average Number of User Ratings: 47220.93548387097
Genre: Utilities
Average Number of User Ratings: 14010.100917431193
Genre: Travel
Average Number of User Ratings: 20216.01785714286
Genre: Shopping
Average Number of User Ratings: 18746.677685950413
Genre: News
Average Number of User Ratings: 15892.724137931034
Genre: Navigation
Average Number of User Ratings: 25972.05
Genre: Lifestyle
Average Number of User Ratings: 8978.308510638299
Genre: Entertainment
Average Number of User Ratings: 10822.961077844311
Genre: Food & Drink
Average Number of User Ratings: 20179.09302325581

From this we can profile an app for the AppStore based on the numbers the app should be profiled inside the categories of Reference, Social & networking and Music, because this are the apps with the highest quantity of ratings, therefore, the more used by users in our app scope. Now, the numbers are not the only thing thata has to be taken in account, because some of this categories are so famous because they have some big apps like Facebook, Twitter, Instagram in Social & Networking and they have some monopoly over this genre, and for Music there are apps like Spotify that get the most users, so this are not optimal genre options for our app because of the competence level. The Reference genre appears to be the most appealing one because it has some famous apps in it but they don't generate a high threat because the difference of usage inside the genre, so an app profiled in this genre can be an option.

For the Play Store there is a column for installs of the app so we can use this to know what are the most popular kinds of app for it. The values in this column are open ended (100+, 1000+, etc) so it is not very precise but for our purposes we don't need complete precision, with this we can mke the analysis. We will assume the numbers as not open ended (100 for 100+, 1000 for 1000+, and so on). But first we need to make some cleaning because we can't transform this value to numbers with the commas and plus signs. For this we will make use of the function replace.

In [23]:
category_ft = create_freqtable(free_google, 1)
for category in category_ft:
    total = 0
    len_category = 0
    for app in free_google:
        category_app = app[1]
        if category_app == category:
            installs = app[5]
            installs = installs.replace('+','')
            installs = installs.replace(',','')
            installs = float(installs)
            total += installs
            len_category += 1
    avg_installs = total/len_category
    print('App Genre: ' + category)
    print('Average installs: ' + str(avg_installs))
    
    

App Genre: ART_AND_DESIGN
Average installs: 1986335.0877192982
App Genre: AUTO_AND_VEHICLES
Average installs: 647317.8170731707
App Genre: BEAUTY
Average installs: 513151.88679245283
App Genre: BOOKS_AND_REFERENCE
Average installs: 8767811.894736841
App Genre: BUSINESS
Average installs: 1712290.1474201474
App Genre: COMICS
Average installs: 817657.2727272727
App Genre: COMMUNICATION
Average installs: 38456119.167247385
App Genre: DATING
Average installs: 854028.8303030303
App Genre: EDUCATION
Average installs: 1833495.145631068
App Genre: ENTERTAINMENT
Average installs: 11640705.88235294
App Genre: EVENTS
Average installs: 253542.22222222222
App Genre: FINANCE
Average installs: 1387692.475609756
App Genre: FOOD_AND_DRINK
Average installs: 1924897.7363636363
App Genre: HEALTH_AND_FITNESS
Average installs: 4188821.9853479853
App Genre: HOUSE_AND_HOME
Average installs: 1331540.5616438356
App Genre: LIBRARIES_AND_DEMO
Average installs: 638503.734939759
App Genre: LIFESTYLE
Average installs

So here we can see that the most popular genre is the communication one but most likely happens the same as in the App Store, it is biased by the influence of big apps like Whatsapp, Facebook, etc. The category of Books and Reference, which is compatible with the chosen genre in the App Store profiling appears to be quiet popular with 8,767,811 installs. Therefore, we can conclude that a good niche for the app in both stores is Books and Reference because it is not biased by the influence of a big competitor and there are many kinds of apps that can be encapsulated in this genres which will give an opportunity to stand out in this market. Maybe a book could be a viable option. 